e2e/cli_test.go

```diff
@@ -319,6 +319,52 @@ func TestLintRuleNamingConventionFromCustomCategory(t *testing.T) {
 	}
 }
 
+func TestAggregatesAreCollectedAndUsed(t *testing.T) {
+	t.Parallel()
+	cwd := must(os.Getwd)
+	basedir := cwd + "/testdata/aggregates"
+
+	t.Run("Zero violations expected", func(t *testing.T) {
+		stdout := bytes.Buffer{}
+		stderr := bytes.Buffer{}
+
+		err := regal(&stdout, &stderr)("lint", "--format", "json", basedir+"/rego", "--rules", basedir+"/rules/custom_rules_using_aggregates.rego")
+
+		if exp, act := 0, ExitStatus(err); exp != act {
+			t.Errorf("expected exit status %d, got %d", exp, act)
+		}
+
+		if exp, act := "", stderr.String(); exp != act {
+			t.Errorf("expected stderr %q, got %q", exp, act)
+		}
+	})
+
+	t.Run("One violation expected", func(t *testing.T) {
+		stdout := bytes.Buffer{}
+		stderr := bytes.Buffer{}
+		// By sending a single file to the command, we skip the aggregates computation, so we expect one violation
```

Comment:
I understand why this happens, but it's a bit of a mind bend to me having test fail by _not_ running aggregate computation. I'd expect `report` rules that need aggregates to just fail out as undefined, likely by referencing `input.aggregate` which would be undefined if no aggregate rules had run?

The way it's done now, there's no way fo a report rule to know if aggregates have been collected or not, or is there?

Reply:
> The way it's done now, there's no way fo a report rule to know if aggregates have been collected or not, or is there?

It is. `input.aggregate` is always defined, which allows devs to check if any aggregate is collected simply via  `count(input.aggregate[_]) > 0` without worrying if it's defined or not. This also allows a kind of typed check, e.g. "check that there are N aggregates of specific type" via `not count([x | x = input.aggregate[_].myType]) == N`

Reply:
Right, but that'll only tell you if any of the aggregate rules has something to report — it's not going to tell you that the aggregate rules were not evaluated at all, which is why I'd prefer it if `input.aggregate` was undefined in that case.

Reply:
> it's not going to tell you that the aggregate rules were not evaluated at all

I did not fully followed this. Maybe some code example? I don't know if by "aggregate rules" you mean rules collecting aggregates (i.e. `aggregate contains entry if`) or the rules evaluating aggregates (i.e. `report contains violation if { .... input.aggregate ... }`

> I'd prefer it if input.aggregate was undefined in that case.

The rationale of merging `not input.aggregates` and `count(input.aggregates[_]) == 0` in to the latter was explained in the [preliminary PR](https://github.com/StyraInc/regal/pull/306#discussion_r1323875777) (in summary: simplify the how one can check aggregates), but that's just my personal preference. If this is a fundamental problem blocking a merge, I don't mind tweaking the PR as needed.
What do you have in mind? Would you like to default on having it `undefined` (if the array is zero), or something different?

Reply:
Sorry, I should have elaborated more on the why, but I was short on time :)

Consider a rule like this: In package `foo` (which could be distributed across multiple files) I need to ensure that there is exactly **one** rule named `bar`. Some pseudo-code below, but bear with me:

```rego
aggregate contains entry if {
    input["package"].name == "foo"
    some rule in input.rules

    rule.head.name == "bar"

    entry := {
        "file" : input.regal.file.name,
        "message": "found rule named 'bar'"
    }
}

report contains violation if {
    count(input.aggregate) != 1
    violation := result.fail(rego.metadata.chain(), {})
}
```

If `input.aggregate` is always an array, this rule will fail even when aggregate collection hasn't happened, like when evaluating only a single file, or pehaps has been disabled by other means. In this case, `count(input.aggregate)` will give me `0`, which is of course not equal to 1. If `input.aggregate` is undefined, evluation will stop at that line, and nothing will be reported.

Reply:
> is always an array, this rule will fail even when aggregate collection hasn't happened, like when evaluating only a single file

😮 This discussion has surfaced a bug: your example above makes sense and (regardless of the original point of the discussion), it seems valid to me to write an aggregate that I want to collect always. The whole point of aggregates was to write business rules I can enforce without concerning how I package the code in multiple `.rego` files. In the current version, a refactor merging all files into one would suddenly cause aggregates not being collected, and thus violations could be missed, violating the principle of least surprise. I think we should:

1) When the docs are added, clearly state that aggregates are collected only if files > 1. Ideally, display a warning if we see aggregate rules unused.
2) Alternatively... be more consistent and remove the optimisation, and always collect aggregates. 

I prefer #2. If you agree I'll make the change. If you prefer #1, no worries (however, I can't commit to making those extra changes soon in follow up PRs due to time constraints).

Now, coming back to the original point, i.e. should be undefined or not:

I think that if we encourage writing rules that attempt to find data more specifically, for example:

```
aggregate contains entry if {
    input["package"].name == "foo"
    some rule in input.rules
    rule.head.name == "bar"
    entry := {
        "bar_rule_found" : input.regal.file.name
    }
}

report contains violation if {
    not count([x | x = input.aggregate[_].bar_rule_found]) == 1
    violation := result.fail(rego.metadata.chain(), {})
}

```
... the logic will work OK for both cases (empty array or undefined).

While I'm still (slightly) inclined to keep it as an empty array, I'm not strong on this one (and I'm here to help the project contributing code and following your lead). I'd be happy to change things. Let me know what changes you'd like and I'll make then soon this week. 

Thanks 🙏

Reply:
> The whole point of aggregates was to write business rules I can enforce without concerning how I package the code in multiple .rego files.

The point of aggregate rules is to provide a way to lint Rego at the level of a _project_, which is something we can't currently do. If your project consists of a single file, there's really no point in using aggregate rules, as there's nothing to aggregate. That case is covered by regular rules. Indeed, this distinction will need to be well documented, and I'd be happy to flesh that out in follow-up PRs.

See the Regal integration that went out in the [Rego Playground](https://play.openpolicyagent.org/) yesterday for an example of single file linting. This will be increasingly common as we start to look into editor integrations later.

> In the current version, a refactor merging all files into one would suddenly cause aggregates not being collected, and thus violations could be missed, violating the principle of least surprise. 

Yes, a refactor merging all files into one would render aggregate rules moot. That's not surprising to me, although I'd be very surprised if someone did that and called it refactoring :)

> I think that if we encourage writing rules that attempt to find data more specifically

Agreed! Even better than encouraging, I think, is a helper function, similar to how we provide `result.fail`. I suggested this [previously](https://github.com/StyraInc/regal/issues/282#issuecomment-1708528606), but much have been said since then, so that it got lost along the way is understandable :) 

Rather than:

```rego
entry := {
    "bar_rule_found" : input.regal.file.name
}
```

We'd provide a `result.aggregate` helper:
```
entry := result.aggregate(rego.metadata.chain(), {"some_other_interesting": "value"})
```

Since we pass `rego.metadata.chain()`, this helper would be able to find the name of the rule from the annotation, and group the aggregates accordingly, i.e. `input.aggregate["my-rule"]`. It would also add `input.regal.file.name` to each aggregate, and anything else we could think might be useful. The second argument would be for the rule author to populate with whatever is useful in the context of their rule.

So yeah 1. would be me preferred choice here:

> 1. When the docs are added, clearly state that aggregates are collected only if files > 1. Ideally, display a warning if we see aggregate rules unused.

And I'd be happy to help fill in some of the blanks in follow-up fixes, like docs, the above mentioned helper function, (optionally) display warnings if aggregate rules are skipped, and so on. Just let me know how you want to proceed 😃

Reply:
> Just let me know how you want to proceed 😃

IIUC, we have two changes:

1) Make aggregates undefined if not collected, or collected for still of zero-length (i.e. normalise to undefined)
2) Add the helper function `result.aggregate(...)`

IMHO, we could leave 2 for later. Initially, Rego authors can put anything they want in the entry. Later, the project can release the helper function to make it easier to auto-insert the rule name, etc.

I'm happy to help with the minimum change you'd accept a merge, leaving any optionals/wish-to-have for a post-merge. If you think we cannot merge this without #2, no worries... I'll commit to add it (not soon because I have too much on my plate recently, but I still want to end what I started). 

The reason I'm leaning towards speed over coding optionals is that the PR has been open for a a while and I have now non-trivial conflicts to merge. I'll work on them while waiting for your answer.